runners/core-java/src/main/java/org/apache/beam/runners/core/DoFnRunner.java

```diff
@@ -52,6 +52,12 @@ void onTimer(
    */
   void finishBundle();
 
+  /**
+   * Calls a {@link DoFn DoFn's} {@link DoFn.OnWindowExpiration @OnWindowExpiration} method and
+   * performs additional task, such as extracts a value saved in a state before garbage collection.
+   */
+  void onWindowExpiration(BoundedWindow window, Instant timestamp, TimeDomain timeDomain);
```

Comment:
What do timestamp and timeDomain mean in this context? 

Also presumably you do want to be able to access the key in onWindowExpiration

Reply:
@reuvenlax timestamp added to keep the time of outputs from OnWindowExpiration. This would be window.maxTimestamp()

timeDomain removed